### ISSUE TITLE
Error building OpenCV 4.3 with QT, OpenGL and CUDA 10.1

### ISSUE THREAD

**Issue by nowhere-girl** [opencv/opencv#17087]
Here's my system information (version)
```
OpenCV 		: 4.3.0
OS		: Ubuntu 18.04 - 64 bit
Compiler	: Native
Graphic card	: Nvidia GTX 950M
Nvidia 		: 440
Cuda  		: Cuda 10.1
```

I configured and generated OpenCV using CMakeGUI, I included opencv_contrib 4.3.0 (for CUDA). My building failed 4 times before (including one with OpenCV 4.2) and just earlier I had **disabled BUILD_EXAMPLES**, **disabled WITH_OPENMP**, and finally came up with this setting :

```
BUILD_TBB:BOOL=1
WITH_TBB:BOOL=1
WITH_OPENGL:BOOL=1
BUILD_PNG:BOOL=1
WITH_PNG:BOOL=1
BUILD_TIFF:BOOL=1
WITH_TIFF:BOOL=1
BUILD_JPEG:BOOL=1
WITH_JPEG:BOOL=1
WITH_CUDA:BOOL=1
OPENCV_DNN_CUDA:BOOL=1
OPENCV_EXTRA_MODULES_PATH:PATH=/home/raisa/Downloads/opencv_contrib-4.3.0/modules
WITH_QT:BOOL=1
OPENCV_ENABLE_NONFREE:BOOL=1
```
Make command with VERBOSE=1, result in this error for opencv_test_viz, but the overall building process didn't stop yet and continue. This is the error :
```
[ 61%] Linking CXX executable ../../bin/opencv_test_viz
cd /home/raisa/Documents/OpenCV/opencv-4.3.0-build/modules/viz && /usr/bin/cmake -E cmake_link_script CMakeFiles/opencv_test_viz.dir/link.txt --verbose=1
/usr/bin/c++     -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winit-self -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -fvisibility-inlines-hidden  -Wno-suggest-override -O3 -DNDEBUG  -DNDEBUG    -Wl,--gc-sections -Wl,--as-needed    CMakeFiles/opencv_test_viz.dir/test/test_common.cpp.o CMakeFiles/opencv_test_viz.dir/test/test_main.cpp.o CMakeFiles/opencv_test_viz.dir/test/test_tutorial2.cpp.o CMakeFiles/opencv_test_viz.dir/test/test_tutorial3.cpp.o CMakeFiles/opencv_test_viz.dir/test/test_viz3d.cpp.o CMakeFiles/opencv_test_viz.dir/test/tests_simple.cpp.o  -o ../../bin/opencv_test_viz  -L/usr/local/cuda-10.1/lib64 -Wl,-rpath,/usr/local/cuda-10.1/lib64:/home/raisa/Documents/OpenCV/opencv-4.3.0-build/lib ../../lib/libopencv_ts.a ../../lib/libopencv_viz.so.4.3.0 ../../lib/libopencv_highgui.so.4.3.0 -ldl -lm -lpthread -lrt /usr/lib/x86_64-linux-gnu/libGL.so /usr/lib/x86_64-linux-gnu/libGLU.so ../../3rdparty/lib/libippiw.a ../../3rdparty/ippicv/ippicv_lnx/icv/lib/intel64/libippicv.a -lcudart_static -lpthread -ldl -lrt -lnppc -lnppial -lnppicc -lnppicom -lnppidei -lnppif -lnppig -lnppim -lnppist -lnppisu -lnppitc -lnpps -lcublas -lcudnn -lcufft -L/usr/local/cuda-10.1/lib64 -L/usr/lib/x86_64-linux-gnu ../../lib/libopencv_videoio.so.4.3.0 ../../lib/libopencv_imgcodecs.so.4.3.0 ../../lib/libopencv_imgproc.so.4.3.0 ../../lib/libopencv_core.so.4.3.0 ../../lib/libopencv_cudev.so.4.3.0 -ldl -lm -lpthread -lrt /usr/lib/x86_64-linux-gnu/libGL.so /usr/lib/x86_64-linux-gnu/libGLU.so -lcudart_static -lpthread -ldl -lm -lpthread -lrt /usr/lib/x86_64-linux-gnu/libGL.so /usr/lib/x86_64-linux-gnu/libGLU.so -lcudart_static -lpthread -lnppc -lnppial -lnppicc -lnppicom -lnppidei -lnppif -lnppig -lnppim -lnppist -lnppisu -lnppitc -lnpps -lcublas -lcudnn -lcufft -Wl,-rpath-link,/home/raisa/Documents/OpenCV/opencv-4.3.0-build/lib 
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFReadDirectory@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFGetFieldDefaulted@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFNumberOfDirectories@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFIsTiled@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFOpen@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFDefaultStripSize@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFReadRGBAImage@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFReadTile@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFSetField@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFWriteScanline@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `_TIFFfree@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFGetField@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFScanlineSize@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFWriteDirectory@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFSetWarningHandler@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFTileSize@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `_TIFFmalloc@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFSetErrorHandler@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFSetDirectory@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFReadScanline@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFClose@LIBTIFF_4.0'
//usr/lib/x86_64-linux-gnu/libvtkIOImage-6.3.so.6.3: undefined reference to `TIFFNumberOfTiles@LIBTIFF_4.0'
collect2: error: ld returned 1 exit status
modules/viz/CMakeFiles/opencv_test_viz.dir/build.make:240: recipe for target 'bin/opencv_test_viz' failed
make[2]: *** [bin/opencv_test_viz] Error 1
make[2]: Leaving directory '/home/raisa/Documents/OpenCV/opencv-4.3.0-build'
CMakeFiles/Makefile2:4373: recipe for target 'modules/viz/CMakeFiles/opencv_test_viz.dir/all' failed
make[1]: *** [modules/viz/CMakeFiles/opencv_test_viz.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....

```
I checked my system and I have libtiff4 (that is needed) and libtiff 5, but still error ?
```
/usr/lib/x86_64-linux-gnu/libtiff.so
/usr/lib/x86_64-linux-gnu/libtiff.so.4
/usr/lib/x86_64-linux-gnu/libtiff.so.4.3.6
/usr/lib/x86_64-linux-gnu/libtiff.so.5
/usr/lib/x86_64-linux-gnu/libtiff.so.5.3.0

```
Then around 64% when building cuda_filters,  process failed completely. In fact, all my failed attempts at building OpenCV, all failed at cuda_filters. Here's the error :

```
[ 64%] Linking CXX shared library ../../lib/libopencv_cudafilters.so
cd /home/raisa/Documents/OpenCV/opencv-4.3.0-build/modules/cudafilters && /usr/bin/cmake -E cmake_link_script CMakeFiles/opencv_cudafilters.dir/link.txt --verbose=1
/usr/bin/c++ -fPIC    -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Winit-self -Wpointer-arith -Wsign-promo -Wuninitialized -Winit-self -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -Wno-long-long -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -msse -msse2 -msse3 -fvisibility=hidden -fvisibility-inlines-hidden -Wno-undef -Wno-missing-declarations -Wno-shadow -O3 -DNDEBUG  -DNDEBUG -Wl,--exclude-libs,libippicv.a -Wl,--exclude-libs,libippiw.a   -Wl,--gc-sections -Wl,--as-needed   -shared -Wl,-soname,libopencv_cudafilters.so.4.3 -o ../../lib/libopencv_cudafilters.so.4.3.0 CMakeFiles/opencv_cudafilters.dir/src/filtering.cpp.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.16sc1.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.16sc3.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.16sc4.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.16uc1.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.16uc3.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.16uc4.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.32fc1.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.32fc3.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.32fc4.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.32sc1.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.32sc3.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.32sc4.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.8uc1.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.8uc3.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_column_filter.8uc4.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_filter2d.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_median_filter.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.16sc1.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.16sc3.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.16sc4.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.16uc1.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.16uc3.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.16uc4.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.32fc1.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.32fc3.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.32fc4.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.32sc1.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.32sc3.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.32sc4.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.8uc1.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.8uc3.cu.o CMakeFiles/cuda_compile_1.dir/src/cuda/cuda_compile_1_generated_row_filter.8uc4.cu.o  -L/usr/local/cuda-10.1/lib64 -Wl,-rpath,/usr/local/cuda-10.1/lib64:/home/raisa/Documents/OpenCV/opencv-4.3.0-build/lib: ../../lib/libopencv_cudaarithm.so.4.3.0 ../../lib/libopencv_imgproc.so.4.3.0 -ldl -lm -lpthread -lrt /usr/lib/x86_64-linux-gnu/libGL.so /usr/lib/x86_64-linux-gnu/libGLU.so ../../3rdparty/lib/libippiw.a ../../3rdparty/ippicv/ippicv_lnx/icv/lib/intel64/libippicv.a -lcudart_static -lpthread -ldl -lrt -lnppc -lnppial -lnppicc -lnppicom -lnppidei -lnppif -lnppig -lnppim -lnppist -lnppisu -lnppitc -lnpps -lcublas -lcudnn -lcufft -L/usr/local/cuda-10.1/lib64 -L/usr/lib/x86_64-linux-gnu -lcudart_static -lpthread -ldl -lrt -lnppc -lnppial -lnppicc -lnppicom -lnppidei -lnppif -lnppig -lnppim -lnppist -lnppisu -lnppitc -lnpps ../../lib/libopencv_core.so.4.3.0 ../../lib/libopencv_cudev.so.4.3.0 -ldl -lm -lpthread -lrt /usr/lib/x86_64-linux-gnu/libGL.so /usr/lib/x86_64-linux-gnu/libGLU.so -lcudart_static -lpthread -lnppc -lnppial -lnppicc -lnppicom -lnppidei -lnppif -lnppig -lnppim -lnppist -lnppisu -lnppitc -lnpps -lcublas -lcudnn -lcufft -ldl -lm -lpthread -lrt /usr/lib/x86_64-linux-gnu/libGL.so /usr/lib/x86_64-linux-gnu/libGLU.so -lcudart_static -lpthread -lnppc -lnppial -lnppicc -lnppicom -lnppidei -lnppif -lnppig -lnppim -lnppist -lnppisu -lnppitc -lnpps -lcublas -lcudnn -lcufft -Wl,-rpath-link,/home/raisa/Documents/OpenCV/opencv-4.3.0-build/lib 
cd /home/raisa/Documents/OpenCV/opencv-4.3.0-build/modules/cudafilters && /usr/bin/cmake -E cmake_symlink_library ../../lib/libopencv_cudafilters.so.4.3.0 ../../lib/libopencv_cudafilters.so.4.3 ../../lib/libopencv_cudafilters.so
make[2]: Leaving directory '/home/raisa/Documents/OpenCV/opencv-4.3.0-build'
[ 64%] Built target opencv_cudafilters
make[1]: Leaving directory '/home/raisa/Documents/OpenCV/opencv-4.3.0-build'
Makefile:162: recipe for target 'all' failed
make: *** [all] Error 2
```

I checked several issues and suggestion, I tried to disable BUILD_EXAMPLES, disable WITH_OPENMP, include EXTRA_MODULES, and make with VERBOSE=1, but still can't seem get it work. Any other suggestion to try ? Thanks in advance.

**Comment by alalek** [comment]
Not sure that this configuration problem can be solved in OpenCV directly.
You should switch to the same libtiff for all components.
You should not use `BUILD_TIFF:BOOL=1` (which triggers own version of libtiff) at least.

> libvtk

Try this workaround: `WITH_WTK=OFF`

---

Build / usage questions should go to Users OpenCV Q/A forum: http://answers.opencv.org

**Comment by nowhere-girl** [comment]
Hi @alalek , thanks. I read than libtiff4 was needed, so I enabled BUILD_TIFF to build its own libtiff, apparently I was wrong. I managed to build and install all of them properly. I didn't opt-out with_vtk, but disabling BUILD_TIFF worked. The problem was I had a whole lot of mess with my libtiff. 
For anyone who's having the same problem, what I did was :

1. Entirely purge (all version of) libtiff and all its dependencies :
	sudo apt-get purge --auto-remove libtiff* (check your own libtiff version)
2. Some version wouldn't get removed properly, so I used force (be careful ! do it at your own risk)
	sudo dpkg --force-all -P libtiff*  (check your own libtiff version)
3. Fix broken packages later with
	sudo apt --fix-broken install  
4. Force install libtiff again with
	sudo apt-get -f install libtiff5-dev

Build and install success, also tested work properly. Things worth mention :
[make with -j] During building, it might get stuck at some point around 90%-99%. It will work, be patient. If not, make without -j.
[make without -j] Take longer, like extremely long. But won't get stuck.
I will close this thread.
